src/daemon/collectd.c

```diff
@@ -83,11 +83,11 @@ static int init_hostname(void) {
 
   str = global_option_get("Hostname");
   if ((str != NULL) && (str[0] != 0)) {
-    sstrncpy(hostname_g, str, hostname_g_size);
+    hostname_set(str);
     return 0;
   }
 
-  if (gethostname(hostname_g, hostname_g_size) != 0) {
+  if (gethostname(hostname_g, sizeof(hostname_g)) != 0) {
```

Comment:
With `hostname_g` being a `char*`, you need to use a temporary buffer here. Sizing is a bit tricky, but the correct way would be:

```c
long hostname_len = sysconf(_SC_HOST_NAME_MAX);
if (hostname_len == -1) {
  hostname_len = NI_MAXHOST;
}
char hostname[hostname_len];
```

Then use `hostname` instead of `hostname_g` for the remainder of the function.

P.S.: If the below `getaddrinfo()` call fails, you'll have to add a `hostname_set(hostname);` to keep the existing behavior of "configured-but-unresolvable is better than no hostname".